Add netlink code and replace networkgateway busybox calls.

tobsan opened the pull request:
All kinds of comments are welcome here! This is work in progress so subject to change quickly.

Commit:
adds libnl test code

Commit:
Netlink code completely replaces busybox calls.

The network gateway now is free of busybox dependencies. This also
replaces inclusion of container.h from gateway.h with the abstract
container interface.

frznlogic commented on common/netlink.h:
```diff
@@ -0,0 +1,64 @@
+#include <linux/rtnetlink.h>
```
This should be documented more properly imho.

tobsan replied:
You mean, in general? I agree! Or do you mean just the rtnetlink header?

frznlogic replied:
The whole file :D

tobsan replied:
Gotcha! Good idea!

frznlogic commented on common/netlink.cpp:
```diff
@@ -0,0 +1,468 @@
+ */
+bool Netlink::get_dump()
+{
+    m_links.clear();
```
indentation in this function...

frznlogic commented on common/netlink.cpp:
```diff
@@ -0,0 +1,468 @@
+		struct msghdr reply;
+		memset(&reply, 0, sizeof(reply));
+
+        struct iovec io = { buf, PAGE_SIZE };
```
indentation.

frznlogic commented on common/netlink.cpp:
```diff
@@ -0,0 +1,468 @@
+        return false;
+    }
+
+	bool bridgeHasGateway = false;
```
indentation

frznlogic commented on common/netlink.cpp:
```diff
@@ -0,0 +1,468 @@
+        msg_setip.pay.ifa_index = ifinfo.ifi_index; // interface (link)
+
+		struct in_addr bcast_addr; // TODO: calculate from ip
+		if (inet_pton(msg_setip.pay.ifa_family, m_broadcast.c_str(), &bcast_addr) == 0) {
```
indentation.

frznlogic commented on common/netlink.cpp:
```diff
@@ -0,0 +1,468 @@
+    if (inet_aton(gateway_address, &gw_addr) == 0) {
+        return false;
+    }
+	add_attribute(set_gw, RTA_GATEWAY, sizeof(gw_addr), &gw_addr);
```
indentation.

frznlogic commented on common/netlink.cpp:
```diff
@@ -0,0 +1,468 @@
+Netlink::Netlink()
+{
+    m_sequence_number = 1;
+    m_ip = "10.0.3.100";
```
These hardcoded IP addresses are really scary imho. Can we do it some other way please?

tobsan replied:
Already done!

frznlogic commented on common/netlink.cpp:
```diff
@@ -0,0 +1,468 @@
+
+template<typename payload> Netlink::netlink_request<payload> Netlink::alloc_msg(const int type, const int flags)
+{
+	netlink_request<payload> request;
```
indentation.

frznlogic commented on common/netlink.cpp:
```diff
@@ -0,0 +1,468 @@
+void Netlink::add_attribute(netlink_request<payload> &req, int type, size_t length, void *data)
+{
+	// First, get a pointer to memory at the end of the payload (pointed out by nlmsg_len)
+	struct rtattr *rta = (struct rtattr *)(((char *) &req) + NLMSG_ALIGN(req.hdr.nlmsg_len));
```
Can this be simplified somehow? It looks really complex because of the typecasts etc, maybe break it down into parts making it simpler to read?

tobsan replied:
That is indeed possible.

frznlogic commented on common/netlink.cpp:
```diff
@@ -0,0 +1,468 @@
+#include <unistd.h> // getpid(), getpagesize()
+#include <string.h> // memcpy, memset etc
+
+Netlink::Netlink()
```
What do you think, should we name the class NetworkManager and the CapabilityManager?

tobsan replied:
Maybe NetlinkManager? But on the other hand, the underlying implementation should not be that interesting to whoever uses this, as long as one can use them both inside and outside containers.

NetworkManager could work, but it has a very GNOMEish ring to it ;)

frznlogic commented on common/netlink.cpp:
```diff
@@ -0,0 +1,468 @@
+
+bool Netlink::setup_netlink()
+{
+    if ((m_fd = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE)) == -1) {
```
What happens if we run setup_netlink multiple times by accident? Do we need some kind of state?

tobsan replied:
Yes, we do need that, thanks for pointing it out!

frznlogic commented on common/netlink.cpp:
```diff
@@ -0,0 +1,468 @@
+					case RTM_NEWLINK:
+                        save_msg<ifinfomsg, LinkInfo>(msg, m_links);
+                        break;
+                    case RTM_NEWADDR:
```
All the breaks feel odd, gather them all up at the end with a default handler perhaps?

frznlogic reviewed (commented):
In general good, some indentation and other comments. I think more documentation of the netlink.h would be really good, possibly some state/state checking of the connection, etc...?

DunderRoffe commented on common/netlink.cpp:
```diff
@@ -0,0 +1,468 @@
+            }
+        }
+
+        // TODO: Make this nicer, please
```
please do

tobsan replied:
Will do!

DunderRoffe reviewed (changes requested):
Should not be put in common as it is not used by agent. Create a separate directory in softwarecontainer for it

Commit:
Docs in the Netlink class + moved class to libSC

Commit:
Changes naming and fixes deallocating data

When storing attributes in the netlink we malloc as much data
as needed and then memcpy from the netlink message to that data.
This commits add a freeing mechanism called during dtor and also
when clearing the cache (before getting a new cache).

DunderRoffe commented on libsoftwarecontainer/src/netlink.cpp:
```diff
@@ -0,0 +1,573 @@
+    return bridgeHasGateway;
+}
+
+int Netlink::readMessage()
```
This method should maybe be split into multiple parts. E.g. One method dealing with link, one method dealing with routes ....  Or maybe it could be solved with some nice template magic ? Thinking mostly of the switch statement.

Commit:
Uses ReturnCode and out-parameters in netlink code

Commit:
SPlit some methods, added further documentation

DunderRoffe commented on libsoftwarecontainer/src/netlink.cpp:
```diff
@@ -0,0 +1,658 @@
+    }
+
+    return ReturnCode::FAILURE;
+}
```
This code does not do what it is supposed to

Commit:
Netlink: adds tests for netlink.

Signed-off-by: Tobias Olausson <tobias.olausson@pelagicore.com>

tobsan commented:
I just updated this PR, please have a look again.

frznlogic commented on libsoftwarecontainer/include/netlink.h:
```diff
@@ -0,0 +1,301 @@
+#include <string>
+#include <vector>
+
+/*
```
/*\* to mark as doxygen documentation.

frznlogic commented on libsoftwarecontainer/include/netlink.h:
```diff
@@ -0,0 +1,301 @@
+         * This clears all the links, addresses and routes. It also has the responsibility to free
+         * all allocated attributes for links, addresses och routes.
+         */
+        void clearCache();
```
Where is the cache created? It might be worth adding a link to it or some note about it.

DunderRoffe commented on libsoftwarecontainer/include/netlink.h:
```diff
@@ -0,0 +1,301 @@
+         * @result ReturnCode::SUCCESS on success, ReturnCode::FAILURE otherwise
+         */
+        template<typename msgtype>
+            ReturnCode getAttributes(const struct nlmsghdr *header, AttributeList &result);
```
Is it possible to use a referance instead of a pointer here ?

tobsan replied:
Yes, I will update it to use that.

DunderRoffe commented on libsoftwarecontainer/src/gateway/networkgateway.cpp:
```diff
@@ -262,64 +261,91 @@ bool NetworkGateway::generateIP()
-    log_debug() << "Attempting to configure eth0 to 'up state'";
-    std::string cmd;
-
     if (!m_interfaceInitialized) {
```
You could negate this and return true inside to lower indentation

tobsan replied:
Good point, will do!

DunderRoffe commented on libsoftwarecontainer/src/netlink.cpp:
```diff
@@ -0,0 +1,559 @@
+
+            // It is ok here for inet_ntop to fail (data could be bad)
+            if (inet_ntop(addressFamily, data, out, sizeof(out))) {
+                if (strcmp(out, needle) == 0) {
```
if (a) if (b) ... == if (a && b)
Maybe it could be rewritten to increase readability ?

tobsan replied:
Good point, will do!

DunderRoffe commented on libsoftwarecontainer/src/netlink.cpp:
```diff
@@ -0,0 +1,559 @@
+        reply.msg_namelen = sizeof(m_kernel);
+
+        len = recvmsg(m_fd, &reply, 0); /* read lots of data */
+        if (len > 0) {
```
Could negate and return to lower indentation

tobsan replied:
Done!

DunderRoffe reviewed (commented):
I guess it is fine either way but here are some suggestions

Commit:
Some updates after comments on PR #60

Signed-off-by: Tobias Olausson <tobias.olausson@pelagicore.com>
